gnovm/pkg/integration/gno.go

```diff
@@ -0,0 +1,103 @@
+// SetupGno prepares the given testscript environment for tests that utilize the gno command.
+// If the `gno` binary doesn't exist, it's built using the `go build` command into the specified buildDir.
+// The function also include the `gno` command into `p.Cmds` to and wrap environment into p.Setup
+// to correctly set up the environment variables needed for the `gno` command.
+func SetupGno(p *testscript.Params, buildDir string) error {
+	// Try to fetch `GNOROOT` from the environment variables
+	gnoroot := os.Getenv("GNOROOT")
+	if gnoroot == "" {
+		// If `GNOROOT` isn't set, determine the root directory of github.com/gnolang/gno
+		goModPath, err := exec.Command("go", "env", "GOMOD").CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("unable to determine gno root directory")
+		}
+
+		gnoroot = filepath.Dir(string(goModPath))
+	}
+
+	if !osm.DirExists(buildDir) {
+		return fmt.Errorf("%q does not exist or is not a directory", buildDir)
+	}
+
+	// Determine the path to the gno binary within the build directory
+	gnoBin := filepath.Join(buildDir, "gno")
+	if _, err := os.Stat(gnoBin); errors.Is(err, os.ErrNotExist) {
+		// Build a fresh gno binary in a temp directory
+		gnoArgsBuilder := []string{"build", "-o", gnoBin}
+
+		// Forward `-covermode` settings if set
+		if coverMode := testing.CoverMode(); coverMode != "" {
+			gnoArgsBuilder = append(gnoArgsBuilder, "-covermode", coverMode)
+		}
+
+		// Append the path to the gno command source
+		gnoArgsBuilder = append(gnoArgsBuilder, filepath.Join(gnoroot, "gnovm", "cmd", "gno"))
+
+		if err = exec.Command("go", gnoArgsBuilder...).Run(); err != nil {
+			return fmt.Errorf("unable to build gno binary: %w", err)
+		}
+	} else if err != nil {
+		// Handle other potential errors from os.Stat
+		return err
+	}
+
+	// Store the original setup scripts for potential wrapping
+	origSetup := p.Setup
+	p.Setup = func(env *testscript.Env) error {
+		// If there's an original setup, execute it
+		if origSetup != nil {
+			if err := origSetup(env); err != nil {
+				return err
+			}
+		}
+
+		// Set the GNOROOT environment variable
+		env.Setenv("GNOROOT", gnoroot)
+
+		// Create a temporary home directory because certain commands require access to $HOME/.cache/go-build
+		home, err := os.MkdirTemp("", "gno")
+		if err != nil {
+			return fmt.Errorf("unable to create temporary home directory: %w", err)
+		}
+		env.Setenv("HOME", home)
+		env.Defer(func() { os.RemoveAll(home) })
+
+		return nil
+	}
+
+	// Initialize cmds map if needed
+	if p.Cmds == nil {
+		p.Cmds = make(map[string]func(ts *testscript.TestScript, neg bool, args []string))
+	}
+
+	// Register the gno command for testscripts
+	p.Cmds["gno"] = func(ts *testscript.TestScript, neg bool, args []string) {
+		err := ts.Exec(gnoBin, args...)
+		if err != nil {
+			ts.Logf("[%v]\n", err)
+			if !neg {
+				ts.Fatalf("unexpected gno command failure")
+			}
+		} else {
+			if neg {
+				ts.Fatalf("unexpected gno command success")
+			}
+		}
+	}
```

Comment:
What do you think about simplifying this as:

```go
err := ts.Exec(gnoBin, args...)
if (err != nil && !neg) || (err == nil && neg) {
    ts.Logf("[%v]\n", err)

	ts.Fatalf("unexpected gno command %s", map[bool]string{true: "success", false: "failure"}[err == nil])
}
```

I don't particularly like this print. 

Maybe we can change it to `unexpected gno command outcome`, and just print the `error == nil` and `neg` values as true / false ?
This way we don't have the print gymnastics and it's pretty straightforward what happened.

ex:
```go
ts.Fatalf("unexpected gno command outcome err=%t neg=%t", err == nil, neg)
```

@ajnavarro what do you think?

Reply:
@zivkovicmilos this part was (certainly) taken from https://github.com/rogpeppe/go-internal/blob/fa6a31e0812c9c062852fbcb34447f2b9e26126d/gotooltest/setup.go#L143-L158 
Im not against changing the way we handle this, but in my opinion I don't think your code block is actually simplify this has the original code is more straightforward.
Your second suggestion otherwise seems to be a good compromise to me :)

Reply:
I 100% agree my suggestion isn't simplified :)

I like what you've done in:
https://github.com/gnolang/gno/pull/1249/commits/d3ec1fd32aaaed1703a228a6b4619bcb05e40ba1

Much better and easier to read 💯